cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -263,23 +281,41 @@ private predicate getLoadedField(LoadInstruction load, Field f, Class c) {
  * Thus, `node1` references an object with a field `f` whose value ends up in
  * `node2`.
  */
-predicate readStep(Node node1, Content f, Node node2) {
+private predicate fieldReadStep(Node node1, FieldContent f, Node node2) {
   exists(LoadInstruction load |
     node2.asInstruction() = load and
     node1.asInstruction() = load.getSourceValueOperand().getAnyDef() and
     exists(Class c |
       c = load.getSourceValueOperand().getAnyDef().getResultType() and
       exists(int startBit, int endBit |
         load.getSourceValueOperand().getUsedInterval(unbindInt(startBit), unbindInt(endBit)) and
-        f.(FieldContent).hasOffset(c, startBit, endBit)
+        f.hasOffset(c, startBit, endBit)
       )
       or
-      getLoadedField(load, f.(FieldContent).getAField(), c) and
-      f.(FieldContent).hasOffset(c, _, _)
+      getLoadedField(load, f.getAField(), c) and
+      f.hasOffset(c, _, _)
     )
   )
 }
 
+private predicate arrayReadStep(Node node1, ArrayContent a, Node node2) {
+  a = TArrayContent() and
+  exists(LoadInstruction load |
+    node1.asInstruction() = load.getSourceValueOperand().getAnyDef() and
```

Comment:
This definition seems overly broad. Can we restrict the address operand to particular syntactic forms like in `arrayStoreStepChi` and ensure there is no overlap with `fieldReadStep`?

Reply:
This was also my main worry when I was CPP-difference, but luckily didn't appear to be a problem. For explicit array and pointer reads (i.e., `x[i]`) we could restrict it by requiring that the source address instruction is a `PointerAdd` instruction. But for an example like:
```cpp
void f(int* s) {
  *s = source();
}

void test() {
  int x;
  f(&x);
  sink(x);
}
```
(where the read step is currently the load of `x` in `sink(x)`) I don't see how to syntactically restrict the predicate.

Reply:
But can you add something like `and not fieldReadStep(node1, _, _)` to ensure there is no overlap?

Reply:
Hm, I was hoping your solution would Just Work, but the following shows that we need to do something else, I think:
```cpp
void taint_a_ptr(int* px) {
  *px = source();
}

void test_outer_with_ptr() {
  Point p;
  f(&p.x);
  sink(p.x);
}
```

Here, the load of `p.x` in `sink(p.x)` actually looks like a `fieldReadStep`, but (as `*px = source()` matches `arrayStoreStepChi`), we actually need an `arrayReadStep` in this case.

Reply:
We discussed this and decided it doesn't need to block this PR.

Reply:
As reported in https://github.com/github/codeql/pull/4298 this is a source of field conflation (which I would've discovered if I had tried to replace `p.x` with `p.y` in the example above 😢). This is fixed by https://github.com/github/codeql/pull/4302.